libmisc/salt.c

```diff
@@ -263,6 +267,78 @@ static /*@observer@*/const char *gensalt_bcrypt (void)
 }
 #endif /* USE_BCRYPT */
 
+#ifdef USE_YESCRYPT
+/* Default cost if not explicitly specified.  */
+#define Y_COST_DEFAULT 8
+/* Minimum cost.  */
+#define Y_COST_MIN 1
+/* Maximum cost.  */
+#define Y_COST_MAX 11
+/*
+ * Return a salt prefix specifying the cost for the YESCRYPT method.
+ */
+static /*@observer@*/const char *YESCRYPT_salt_cost (/*@null@*/long *prefered_cost)
+{
+	static char cost_prefix[5];
+        long cost;
+
+        if (NULL == prefered_cost) {
+	    cost = getdef_num ("YESCRYPT_COST_FACTOR", Y_COST_DEFAULT);
+        } else {
+            cost = *prefered_cost;
+        }
+
+        if (cost < Y_COST_MIN) {
+            cost = Y_COST_MIN;
+        }
+        if (cost > Y_COST_MAX) {
+            cost = Y_COST_MAX;
+        }
+
+        cost_prefix[0] = 'j';
+        if (cost < 3) {
+            cost_prefix[1] = 0x36 + cost;
+        } else if (cost < 6) {
+            cost_prefix[1] = 0x34 + cost;
+        } else {
+            cost_prefix[1] = 0x3b + cost;
+        }
+        cost_prefix[2] = cost >= 3 ? 'T' : '5';
+        cost_prefix[3] = '$';
+        cost_prefix[4] = 0;
```

Comment:
Oh, this is a rather compact way to mimic (I guess) libxcrypt's translation of a single cost parameter to yescrypt's multiple ones, which it'd then encode with `yescrypt_encode_params_r()`. I didn't verify it, but I assume the resulting strings were checked to match what libxcrypt produces? If so, that's fine.

Reply:
Exactly. At first I tried to actually implement a proper parameter generation. Unfortunately, I found no documentation and very few comments in the code. I did not spent too much time on it, but I think it involves bit masks and non-standard base64 encoding. Instead of reading some foreign code for days and re-implementing it, I generated the 11 possible values using libxcrypt's `crypt_gensalt`.

Reply:
I accept your criticism here: I should write a specification of the encoding, document the rationale behind the choices, and add some source code comments. The compact encoding of numeric parameters is actually specific to yescrypt, and needs comments.